platform/linux-generic/odp_crypto.c

```diff
@@ -308,6 +317,148 @@ void packet_hmac(odp_packet_t pkt,
 	HMAC_Final(ctx, hash, NULL);
 }
 
+static void do_pad_xor(uint8_t *out, const uint8_t *in, int len)
+{
+	int pos = 0;
+
+	for (pos = 1; pos <= 16; pos++, in++, out++) {
+		if (pos <= len)
+			*out ^= *in;
+		if (pos > len) {
+			*out ^= 0x80;
+			break;
+		}
+	}
+}
+
+static void xor_block(uint32_t *res, const uint32_t *op)
+{
+	res[0] ^= op[0];
+	res[1] ^= op[1];
+	res[2] ^= op[2];
+	res[3] ^= op[3];
+}
+
+static
+void packet_aes_xcbc_mac(odp_packet_t pkt,
+		 const odp_crypto_packet_op_param_t *param,
+		 odp_crypto_generic_session_t *session,
+		 uint8_t *hash)
+{
+	uint32_t e[4] = {0, 0, 0, 0};
+	uint32_t offset = param->auth_range.offset;
+	uint32_t len   = param->auth_range.length;
+	uint32_t seglen = 0;;
+	int dummy_len = 0;
+	EVP_CIPHER_CTX *ctx;
+	void *mapaddr;
+	uint8_t *data;
+	
+	ODP_ASSERT(offset + len <= odp_packet_len(pkt));
+
+	ctx = EVP_CIPHER_CTX_new();
+	EVP_EncryptInit_ex(ctx, session->auth.evp_cipher, NULL, session->auth.key, NULL);
+	mapaddr = odp_packet_offset(pkt, offset, &seglen, NULL);
+	data = (uint8_t *)mapaddr;
+	for (; len > AES_BLOCK_SIZE ; len -= AES_BLOCK_SIZE) {
+		xor_block(e, (const uint32_t *) data);
+		EVP_EncryptUpdate(ctx, (uint8_t *)e, &dummy_len, (uint8_t *)e, sizeof(e));
+		data += AES_BLOCK_SIZE;
+	}
+	do_pad_xor((uint8_t *)e, data, len);
+	if (len == AES_BLOCK_SIZE)
+		xor_block(e, (const uint32_t *)(session->auth.key + 16));
```

Comment:
16 is magic number. I'd include use `EVP_CIPHER_key_length()` instead of just 16 here.